// cspell-lib/dist/esm/util/url.js
import path from 'path';
import { fileURLToPath, pathToFileURL } from 'url';
import { srcDirectory } from '../../lib-cjs/pkg-info.cjs';
const isUrlRegExp = /^(?:[\w][\w-]+:\/|data:|untitled:)/i;
/**
 * Convert a URL into a string. If it is a file URL, convert it to a path.
 * @param url - URL
 * @returns path or href
 */
export function toFilePathOrHref(url) {
    return fileURLOrPathToPath(url);
}
/**
 * This is a URL that can be used for searching for modules.
 * @returns URL for the source directory
 */
export function getSourceDirectoryUrl() {
    const srcDirectoryURL = pathToFileURL(path.join(srcDirectory, '/'));
    return srcDirectoryURL;
}
/**
 * @param path - path to convert to a URL
 * @param relativeTo - URL to resolve the path against or the current working directory.
 * @returns a URL
 */
export function relativeTo(path, relativeTo) {
    return new URL(normalizePathSlashesForUrl(path), relativeTo || cwdURL());
}
export function cwdURL() {
    return pathToFileURL('./');
}
export function resolveFileWithURL(file, relativeToURL) {
    if (file instanceof URL)
        return file;
    if (isURLLike(file))
        return toURL(file);
    const isRelativeToFile = isFileURL(relativeToURL);
    if (isRelativeToFile && path.isAbsolute(file)) {
        return pathToFileURL(file);
    }
    if (isRelativeToFile) {
        const rootURL = new URL('.', relativeToURL);
        const root = fileURLToPath(rootURL);
        const suffix = file === '.' || file == '..' || file.endsWith('/') || file.endsWith(path.sep) ? '/' : '';
        const filePath = path.resolve(root, file);
        return pathToFileURL(filePath + suffix);
    }
    return relativeTo(file, relativeToURL);
}
export function normalizePathSlashesForUrl(filePath, sep = /[/\\]/g) {
    return filePath
        .replace(/^([a-z]:)/i, '/$1')
        .split(sep)
        .join('/');
}
export function toFileUrl(file) {
    if (file instanceof URL)
        return file;
    return resolveFileWithURL(file, cwdURL());
}
export function toFileDirUrl(dir) {
    return addTrailingSlash(toFileUrl(dir));
}
export function addTrailingSlash(url) {
    if (url.pathname.endsWith('/'))
        return url;
    const urlWithSlash = new URL(url.href);
    urlWithSlash.pathname += '/';
    return urlWithSlash;
}
export function toURL(href, relativeTo) {
    return href instanceof URL ? href : new URL(href, relativeTo);
}
export function fileURLOrPathToPath(filenameOrURL) {
    return isFileURL(filenameOrURL) ? toFilePath(filenameOrURL) : filenameOrURL.toString();
}
function toFilePath(url) {
    return windowsDriveLetterToUpper(fileURLToPath(url));
}
export function isURLLike(url) {
    return url instanceof URL || isUrlRegExp.test(url);
}
export function isFileURL(url) {
    return isUrlWithProtocol(url, 'file');
}
export function isDataURL(url) {
    return isUrlWithProtocol(url, 'data');
}
function isUrlWithProtocol(url, protocol) {
    protocol = protocol.endsWith(':') ? protocol : protocol + ':';
    return url instanceof URL ? url.protocol === protocol : url.startsWith(protocol);
}
export function windowsDriveLetterToUpper(absoluteFilePath) {
    return absoluteFilePath.replace(/^([a-z]):\\/, (s) => s.toUpperCase());
}
