// cspell/dist/esm/lint/LintRequest.mjs
import * as path from 'path';
import { calcExcludeGlobInfo } from '../util/glob.mjs';
const defaultContextRange = 20;
export class LintRequest {
    fileGlobs;
    options;
    reporter;
    uniqueFilter;
    locale;
    configFile;
    excludes;
    root;
    showContext;
    enableGlobDot;
    fileLists;
    constructor(fileGlobs, options, reporter) {
        this.fileGlobs = fileGlobs;
        this.options = options;
        this.reporter = reporter;
        this.root = path.resolve(options.root || process.cwd());
        this.configFile = options.config;
        this.excludes = calcExcludeGlobInfo(this.root, options.exclude);
        this.locale = options.locale || '';
        this.enableGlobDot = options.dot;
        // this.uniqueFilter = options.unique ? util.uniqueFilterFnGenerator((issue: Issue) => issue.text) : () => true;
        this.uniqueFilter = () => true;
        this.showContext =
            options.showContext === true ? defaultContextRange : options.showContext ? options.showContext : 0;
        this.fileLists = (options.fileList ?? options.fileLists) || [];
    }
}
