// cspell-trie-lib/dist/lib/TrieBlob/FastTrieBlobIRoot.js
const EmptyKeys = Object.freeze([]);
const EmptyNodes = Object.freeze([]);
class FastTrieBlobINode {
    trie;
    nodeIdx;
    id;
    size;
    node;
    eow;
    charToIdx;
    _keys;
    constructor(trie, nodeIdx) {
        this.trie = trie;
        this.nodeIdx = nodeIdx;
        const node = trie.nodes[nodeIdx];
        this.node = node;
        const value = node[0];
        this.eow = !!(value & trie.NodeMaskEOW);
        this.size = node.length - 1;
        this.id = nodeIdx;
    }
    keys() {
        return (this._keys ??= this.calcKeys());
    }
    /** get keys to children */
    calcKeys() {
        if (!this.size)
            return EmptyKeys;
        const NodeMaskChildCharIndex = this.trie.NodeMaskChildCharIndex;
        const charIndex = this.trie.charIndex;
        const keys = Array(this.size);
        const len = this.size;
        const node = this.trie.nodes[this.nodeIdx];
        for (let i = 0; i < len; ++i) {
            const entry = node[i + 1];
            const charIdx = entry & NodeMaskChildCharIndex;
            keys[i] = charIndex[charIdx];
        }
        return Object.freeze(keys);
    }
    values() {
        if (!this.size)
            return EmptyNodes;
        const nodes = Array(this.size);
        for (let i = 0; i < this.size; ++i) {
            nodes[i] = this.child(i);
        }
        return nodes;
    }
    entries() {
        const keys = this.keys();
        const values = this.values();
        const len = keys.length;
        const entries = Array(len);
        for (let i = 0; i < len; ++i) {
            entries[i] = [keys[i], values[i]];
        }
        return entries;
    }
    /** get child ITrieNode */
    get(char) {
        const idx = this.getCharToIdxMap()[char];
        if (idx === undefined)
            return undefined;
        return this.child(idx);
    }
    has(char) {
        const idx = this.getCharToIdxMap()[char];
        return idx !== undefined;
    }
    hasChildren() {
        return this.size > 0;
    }
    child(keyIdx) {
        const node = this.trie.nodes[this.nodeIdx];
        const nodeIdx = node[keyIdx + 1] >>> this.trie.NodeChildRefShift;
        return new FastTrieBlobINode(this.trie, nodeIdx);
    }
    getCharToIdxMap() {
        const m = this.charToIdx;
        if (m)
            return m;
        const map = Object.create(null);
        const keys = this.keys();
        for (let i = 0; i < keys.length; ++i) {
            map[keys[i]] = i;
        }
        this.charToIdx = map;
        return map;
    }
}
export class FastTrieBlobIRoot extends FastTrieBlobINode {
    info;
    constructor(trie, nodeIdx, info) {
        super(trie, nodeIdx);
        this.info = info;
    }
    resolveId(id) {
        return new FastTrieBlobINode(this.trie, id);
    }
}
